### DISCUSSION TITLE
Is it possible to get the implicit store?

### DISCUSSION THREAD

**Discussion by matthew-dean** [pmndrs/jotai#1451] [General]
Still more store questions, sorry!

Today's question. So, I understand that you an do `createStore()` (currently `unstable_createStore`) and my understanding is that you can set that store for the tree using `<Provider>`.

My question is, for provider-less mode, is it possible to just "get" that "provider-less" store? Like the implicit store that Jotai will create? Such that I don't have to require the use of `<Provider>` for an end-user where I am using a Jotai store to get/set values in utility functions?

Something like:
```jsx
import { getStore, atom } from 'jotai'

const atomWhoseValueHasBeenSetByAComponent = atom(0)
const store = getStore()

// ...later, after my React tree has done things...
const foo = store.get(atomWhoseValueHasBeenSetByAComponent)
```

**Comment by dai-shi** [comment]
You are welcome to have many questions. 
The default store for provider less mode is hidden. It's technically possible to expose it, but that's why provider is provided to replace the default store. Isn't it too much trouble to put one provider component at your root of the tree?

**Reply by matthew-dean** [reply]
The problem is that it makes it hard to use general-purpose functions that rely on `useAtom`. That is, it becomes an implicit dependency. It's that it's not that much "trouble", it just has me writing code like this:
```tsx
export const featureFlagsAtom = atomWithStorage<LDFlagSet>('featureFlags', {})

/**
 * We do this instead of `useAtom` so that we make sure we're
 * using the same store for setting / getting values.
 */
export const useFeatureFlags = () => {
  const [flags, setFlags] = useState(store.get(featureFlagsAtom))
  useEffect(() => {
    const unsub = store.sub(featureFlagsAtom, () => {
      setFlags(store.get(featureFlagsAtom))
    })
    return unsub
  })
  return flags
}

export const setupFeatureFlags = (clientId: string, user?: LDUser) => {
  const client = LDClient.initialize(clientId, user ?? { anonymous: true }, {
    sendEventsOnlyForVariation: true
  })
  client.on('initialized', () => {
    store.set(featureFlagsAtom, client.allFlags())
  })
  client.on('change', () => {
    store.set(featureFlagsAtom, client.allFlags())
  })
  return { client }
}
```
To walk you through it, basically, `useFeatureFlags` should be a minimal abstraction. BUT, I can't use `useAtom` because I'm setting, via code, the value of `featureFlagsAtom` in `store`, and this is done outside the context of a component. UNLESS I also force the end-user to also use the exported store as their main and only store and set it with `<Provider>`.

I could either:
1. force `setupFeatureFlags` to be used only in a component, negating the requirement of `createStore`, and then refactor `useFeatureFlags` to not use `store`
2. force the use of `Provider` to the end-user and require to use the exported store

So my point is that not being able to get the default store leaves me minimal options. As soon as I create a store, there's additional cognitive overhead. In reality, I don't _really_ want to create a store at all; I just want to programmatically get / update the default store. And, actually, my guess is that this is the actual need (`getStore`), and not `createStore`. Very few people should need to create stores IMO. They just need / want programmatic access (regardless if the React tree has been mounted or not).

**Reply by matthew-dean** [reply]
> And, actually, my guess is that this is the actual need (getStore), and not createStore. Very few people should need to create stores IMO

Here's [an example issue](https://github.com/pmndrs/jotai/issues/864) to illustrate my point. The author of that issue is wanting to write the value to the current (default) store. I'm just skeptical that the need for `createStore()` exceeds the need for `getStore()`. The simplest use case is `getStore()` (or named `getDefaultStore()`)

**Reply by matthew-dean** [reply]
To be clear, I would re-write this as:
```tsx
import { getDefaultStore, useAtom } from 'jotai'
import { atomWithStorage } from 'jotai/utils'

const store = getDefaultStore()

export const featureFlagsAtom = atomWithStorage<LDFlagSet>('featureFlags', {})

/**
 * We can use `useAtom`!
 */
export const useFeatureFlags = () => {
  const [flags] = useAtom(featureFlagsAtom)
  return flags
}

export const setupFeatureFlags = (clientId: string, user?: LDUser) => {
  const client = LDClient.initialize(clientId, user ?? { anonymous: true }, {
    sendEventsOnlyForVariation: true
  })
  client.on('initialized', () => {
    store.set(featureFlagsAtom, client.allFlags())
  })
  client.on('change', () => {
    store.set(featureFlagsAtom, client.allFlags())
  })
  return { client }
}
```

**Reply by matthew-dean** [reply]
One thing to clarify and may make this easier to understand -- This is a code contract between two teams. One building shared components and utilities, and another building apps. It's an easier contract if the app team is not forced to add things to their tree and can just import / use utilities and hooks as-is. It's not the _end of the world_ if they had to add a Provider; we would just prefer it wasn't necessary, and that it we wouldn't add Provider bloat to each app.

**Reply by dai-shi** [reply]
Notice useAtom is contextual. I don't think exposing default store solves your issue completely. It can access to a different store. 

To mitigate misusing a different store, what I can think of is to add a special atom and check its value in advance in a custom hook.

**Reply by matthew-dean** [reply]
@dai-shi 

> Notice useAtom is contextual. I don't think exposing default store solves your issue completely. 

I get that it's contextual. My point is that if I could get the default store, I would _only_ use that store, which means all functions would use that as the store context without an explicit `<Provider>`. I wouldn't use `createStore()`.

**Reply by dai-shi** [reply]
But, your teammates can break the rule and use Provide somewhere, right? It then doesn't read from the default store.
